### DISCUSSION TITLE
「ジャッカルとサイドキック」問題をどうにかしたい

### DISCUSSION THREAD

**Discussion by yukieiji** [yukieiji/ExtremeRoles#47] [Ideas]
現在、EXRolesで15人バランス向けルールを考えた時に「ジャッカルとサイドキック」はほぼ確定で採用される。そのような現状は以下の理由でよろしくない
- 理由
 1. サイドキックを作った時の処理を考え続ける必要がある
    - この役職は上書きされるのか？上書きされないのか？とかを考える必要がある(開発者もプレイしてる人達も)、それをプレイヤーに任せるなら新しいオプションを作り続ける必要がある
      -  例1：「アサシンとマーリン」でマーリンがサイドキックにされた時の処理
      -  例2：「パン屋」がサイドキックにされた時の処理
         - 両方共オプションで切り替えることは技術的には可能であるが、それを実装するコストやリスク(不具合発生確率やコードの複雑性)に対するメリットが少なすぎる(新しく役職を作ったほうが数倍楽)
            - どれくらいのリスクがあるかというとこのチケット発行時のExRのリリース51件中「ジャッカルとサイドキック」周りの不具合修正が11件、改善やオプション込みだと18件もある(ここに新しい何かを入れればその件数は増えていくことは目に見えてわかる)
    - サーヴァントの場合はニュートラル以外必ずリセットされない、ドールの場合は必ずリセットすると言うコンセプトがある
 3. 15人の第三陣営はほぼ「ジャッカルとサイドキック」になるので新しい役職のバランスを考える時にそれが前提になる
    - 「「ジャッカルとサイドキック」だからこの提案されてる役職はこうだよね？」が続くと役職のバランス調整を行ったら「ジャッカルとサイドキック」側も調整する必要が出てくる(プログラミング的に言えば依存関係が複雑だから、AをいじったらBもいじらないといけない)
    - バランス的にも役職はその役職が届かない範囲は基本的にいじらないし依存しない方がいい
 4. 役職の飽き
    - 15人だとほぼ「ジャッカルとサイドキック」なのでニュートラル引いた時のワクワク感は時間が経てば経つほど薄れていき、喪失感は増えていく
    - 戦術やプレイングも人が同じであれば同じであるというメタ読みが通り始める

上記問題「「ジャッカルとサイドキック」問題」をどうにかしたいので「ジャッカルとサイドキック」に引けを取らない様なニュートラル役職を作りたいし実装したい
この問題に対してのみんなの意見(こうすればいい、そんなことはない、etc...)や役職案、役職のコンセプト(まとまってなくてもOK！！)、こんな能力をもつニュートラル役職が欲しいとかを聞きたい
issuesじゃなくDiscussionなので実装とかバランス云々はあんまり気にしないし、気にしてたら言って下さい

参考資料：EXRolesのニュートラル達の立ち位置的なもの
![neutrals](https://user-images.githubusercontent.com/40425798/168340591-22f2669c-9760-4c9f-b8aa-a15206b7677b.PNG)

2022/05/14T21:19追記：
- 「ジャッカルとかサイドキック」を排除したいわけではないですし、「ジャッカルとサイドキック」に対して今後一切の調整を行わない訳では無いです

2022/11/04T1:28追記：
- 少し説明を更新

**Comment by melonjagaimo** [comment]
貴族&奴隷なんてどうですかね
・貴族はゲーム開始時に何人かのプレイヤーを奴隷にする（インポスター陣営は必ず1人だけ奴隷になる）
・奴隷は元々の役職を失わず、自身を所有する貴族が生きている限り絶対に勝てない。それどころか生きている奴隷がいる陣営が勝つと、その勝利を貴族が丸ごと奪って貴族が単独勝利してしまう
・貴族が通常のキルをされる時、自身が所有し生きている奴隷を1人身代わりに死なせる（身代わりになった奴隷はその場で倒れる）
・ただし奴隷が、自身を所有する貴族をキルしようとした場合は例外的に貴族を一方的にキルできる
・奴隷は奴隷であることを把握しているが誰が自身を所有する貴族かは把握しない。
・貴族が追放や旅立ちで死亡した場合、またはサイドキックにされた場合、その貴族が所有する奴隷は全員解放される。奴隷が解放されたことは奴隷に通知され、解放された奴隷は元々の役職として勝利できる
・奴隷がサイドキックにされた場合、その奴隷は解放される
貴族の勝利条件
①生存者が貴族と、その貴族の所有する奴隷のみの場合。貴族の単独勝利
②貴族が生存している状態で、その貴族が所有する奴隷のいる陣営が勝利する場合。この時、貴族は勝利を奪って単独勝利
奴隷の勝利条件
自身または自身を所有する貴族が死亡している状態で、元々の役職の陣営が勝つ時。

**Reply by yukieiji** [reply]
ゲーム開始時から「奴隷」が「奴隷」であると認識できるようにするってなると他の役職の処理に割り込みをかける必要が出てきてしまいます(理由2の「役職はその役職が届かない範囲は基本的にいじらないし依存しない方がいい」が未解決になってしまう)
技術的な観点の話が出てきてしまいますが、役職アサインのロジックに「貴族と奴隷」用の特殊な専用処理が必要になると思います(IRoleSpecialSetUpでの追加処理だと速度的な問題で間に合わないはず)
認識できる奴隷自体はMultiAssignRoleBaseで実装できるかなって一瞬考えましたけどラバーズ+〇〇とかが奴隷になると実装的に不具合が生じますね

**Reply by melonjagaimo** [reply]
そうですか・・・。
個人的には「この役職が生きている限り、クルー・インポスター双方勝てないのでどうにか死なせたい役職」を実装すれば解決の糸口になると思ってます
それこそ本家人狼で言うところの妖狐のように

**Reply by yukieiji** [reply]
「貴族と奴隷」の仕様的には問題はないとは思います(バランスはさておいてとして)
「認識できる奴隷」って部分で問題がある感じです

「妖狐」に関しては #2 で議論されていまして、能力さえ決まれば実装をしようと考えております(意見等がある場合はチケットにお願いします)

**Reply by melonjagaimo** [reply]
奴隷が奴隷だと認識できない、ですか
それもよさそうですね。突然クルーが倒れたら議論が混乱しそうですし（シェリフがいるとそのシェリフが自爆したように見える）

**Reply by yukieiji** [reply]
自分の中で案が固まり、新しくチケットを出して頂ければそちらでバランス等も含めて議論させていただきます

**Comment by Inumiya** [comment]
ジャッカルがなぜこんなにも使われているのか、まずはその要因を列挙して、その後それらの要因を満たす役職を考える方が良いかと想います

- 役職の説明が簡単
- キル能力を持っている
  - インポスターと食い合う可能性があり、インポスターの一方的な試合になるのを防止できる
  - インポスターに「ジャッカルは誰なのか」という推理要素を追加できる
- 能力を色々な使い方が出来る
  - サイドキックを温存するか、即使うか
  - ここに個人差が出て面白い
- 自分が死んでもまだ勝ち目がある
  - 他の第３陣営は、自陣営が自分一人だと辛いが、増やすとキルペースが上がりすぎる
  - ジャッカル・サイドキックはこの欠点を克服している

まず思い浮かぶのはこれくらいでしょうか

**Reply by yukieiji** [reply]
> ジャッカルがなぜこんなにも使われているのか、まずはその要因を列挙して、その後それらの要因を満たす役職を考える方が良いかと想います

ありがとうございます。たしかにそれも重要ですね

私自身ジャッカルが使われている理由は以下だと思います
1. 第三陣営の中でも勝率が良い感じになってる
   - 勝ちすぎず負け過ぎない感じ
2. キルボタンによる人数減少と能力で誰でも自陣営に引き込むという試合を役職の人がコントロール出来る要因を2つ持っている
   - これが多分1の理由であると思われる
   - キル自体にゲーム的なジレンマがあるし、能力の使い所にもゲーム的なジレンマがあるのでそれ自体も良いスパイスだと思います
3. インポスターとクルー両方に程よい感じのデメリットが有る
   - 戌宮さんの2つ目の理由とほぼ同じだと思います
   - 多人数ルールでアリスだとデメリットが強すぎる、ミッショナリーだとデメリットが弱すぎる(クルー側のデメリットが弱いのかな？)

**Reply by Inumiya** [reply]
> キルボタンによる人数減少と能力で誰でも自陣営に引き込むという試合を役職の人がコントロール出来る要因を2つ持っている

ここが新役職で満たせれば良いのですが、難しいですね
キル能力を持たせるのは良いとして、他陣営を自陣営に引き込む以外の方法で盤面コントロールを行う方法が思いつかないです

**Reply by yukieiji** [reply]
> キル能力を持たせるのは良いとして、他陣営を自陣営に引き込む以外の方法で盤面コントロールを行う方法が思いつかないです

盤面コントロールの方法には幾つかあると思います
- プレイヤーの位置をランダムに入れ替える
- ありとあらゆるボタンを封じる
- タスクの難易度を上げる
- ...etc

参考までに今のEXRolesのニュートラルが持ってるゲームコントロール能力を以下に記載します
- アリス：「破壊工作」によるタスク巻き戻し、インポスターにキルされるとゲーム強制終了、キルボタン、サボタージュ
- タスクマスター：サボタージュ、(新規実装予定能力)
- ミッショナリー：特殊遠隔キルによる盤面かき乱し
- ジェスター：キル能力強制発動による吊られやすさの向上と邪魔なキル陣営の排除、サボタージュ、吊られるとゲーム強制終了
- ヤンデレ：大きく増減するキルクールによって盤面を大きくひっくり返せる
- ヴィジランテ：「状態」に応じて付く陣営が変わる、自分である程度コントロール出来る

**Reply by Inumiya** [reply]
ありがとうございます

ジャッカルの代わりに使われる役職なら、ジャッカルの欠点も克服できていると良いかなと思いジャッカルの欠点をTwitterで募集してみました

- サイドキック作る前に死ぬとヒマ
- サイドキック出来ない役職がいる
- サイドキック以外に特異な能力がない

辺りが出てきました
これらを考慮して、盤面コントロール手段が複数あるという要素を入れると、
「生前はキル能力で盤面コントロールして、死後は別の能力で盤面コントロールする（自陣営が全滅後もうまくやると勝利条件を満たせる）」
という役職が良いかなと思いました

**Reply by yukieiji** [reply]
> 「生前はキル能力で盤面コントロールして、死後は別の能力で盤面コントロールする（自陣営が全滅後もうまくやると勝利条件を満たせる）」

死後の能力に関しては幽霊役職の仕様が決まっていないので何とも言えない部分があります
幽霊役職と生きている時の役職は一応別扱いになる予定なのですがそれは問題ない感じでしょうか(ある役職からしかアサインされない幽霊役職みたいなことは出来るようにはなると思われますが)

**Reply by Inumiya** [reply]
生前役職Xからしか死後役職Yにはなれない仕様がいいかと思います
あとXとYはコンビ役職にして、Xがアサインされた場合必ずXは死後Yになる仕様でお願いしたいです
Yがアサインされる場合のXと、Yがアサインされない場合のXの２パターンを考えてバランス調整するのは難しそうなので

**Reply by yukieiji** [reply]
> あとXとYはコンビ役職にして、Xがアサインされた場合必ずXは死後Yになる仕様でお願いしたいです

今提案されている仕様と私が考えている仕様的にそれが出来るかどうかってのは微妙な感じです

もしそう言った仕様が欲しい場合は #46 辺りに投げて頂ければ

**Reply by Inumiya** [reply]
了解しました

**Comment by kisaRagi05** [comment]
ジャッカルが使われる要因の一つ下記の内容があると思います。

- クルー陣営の役職COが確定白要素になるのを防ぐ
  - サイドキックがあるから、役職COも確実に白とはならない
  - クルー陣営の役職CO対策が他の第三陣営に無いので非常に採用されやすい

なので「ジャッカル＆サイドキック」の代わりになりうる第三陣営は
クルー陣営の役職COの信頼度を下げる要素が必要となると思います

そこでですが、TORのイビル○○のようにクルーの役職にキルボタン追加は可能でしょうか？
つまり白役職を持ちキル能力を持つ第三陣営が分かりやすくジャッカルの代わりになりえそうかなと思いました

白役職を持っているのにキルボタンを持ちクルーの勝利では無く自分の陣営の勝利を目指す裏切者
仮称として「トレイター」とします

これで全員役職持ちでクルー陣営が全員役職COしても、その中にトレイターが含まれる場合もあるとして
役職の無い普通のクルーだと言ってる人だけを怪しむとかも無くなると思います

**Reply by yukieiji** [reply]
> そこでですが、TORのイビル○○のようにクルーの役職にキルボタン追加は可能でしょうか？

一応、イビルサポーターと言うものが居るためそう言ったイビルとナイス両対応はできると思います

> 白役職を持っているのにキルボタンを持ちクルーの勝利では無く自分の陣営の勝利を目指す裏切者
仮称として「トレイター」とします

キルボタン云々は置いておいて、最初クルーでアサインされるけど初期化処理中にニュートラルしてしまえばそういった処理は可能だと思います(クルーの能力がくっついて来ますけど、トレイター+TMみたいな感じ？)

PS.奪った後の初期化中に役職を葬り去れば、クルー能力がくっついてくる問題も解決できるなと今考えました

**Reply by kisaRagi05** [reply]
> キルボタン云々は置いておいて、最初クルーでアサインされるけど初期化処理中にニュートラルしてしまえばそういった処理は可能だと思います(クルーの能力がくっついて来ますけど、トレイター+TMみたいな感じ？)

そんな感じになります
「ジャッカル＆サイドキック」と同じくサボタージュが出来ない変わりにクルーの能力が使える感じとなり
「トレイター+TM」とかなら能力「タイムシールド」も使えてキルも出来るとなります（能力証明による確定白対策）
これがあればトレイターTMがフェンサーを噛み、フェンサーが反撃キルしてタイムシールド発動とかの
今までにない殴り合いが見れるかもと様々な可能性を持ってるはずです

> PS.奪った後の初期化中に役職を葬り去れば、クルー能力がくっついてくる問題も解決できるなと今考えました

技術的に難しいようであればそれも良いかと思いますが
クルー役職の名前だけで実際の能力無しだとサイドキックと同じとなるので
何か特異な能力を追加が欲しいですね

**Reply by yukieiji** [reply]
> 技術的に難しいようであればそれも良いかと思いますが

いや、ただ単にトレイター+シェリフみたいなことになった場合、クルーを切れない(切った瞬間爆発)という感じになってしまうことに気づいたからです
基本的にクルー役職はクルー有利やクルー前提の能力になっているため変なことを考えるよりかは消し去ったほうがいいかなって思いました

まぁここらへんは役職の内部仕様に近いと思うのでissuesのチケットが出てから考えてもいいと思います

**Reply by kisaRagi05** [reply]
確かにそれ以上の詳細はissuesを作成してそこでが良さそうですね
内容をまとめて改めてお送りします

**Comment by wilfrem** [comment]
整理すると
問題1: 「サイドキックがバグの元で対応コストが高いので、よい形で"排除"したい」
問題2: 「ジャッカルの採用率が高いのを何とかしたい」

そのための対策として
「ジャッカルよりも魅力的なニュートラル役職を作りたい」

これにより二つの問題を一度に解決したい！

で合ってますでしょうか？


そして、そうすると、「やらない」「やりたくない」こととして

- ジャッカル/サイドキックをNerfする
- ジャッカルの勧誘周りを対応コストが低い形に修正する
- 新たな勧誘能力を持ったニュートラル役職の作成

あたりが「やらない」「やりたくない」事項になりますか？

**Reply by yukieiji** [reply]
> 問題1: 「サイドキックがバグの元で対応コストが高いので、よい形で"排除"したい」

別に排除したいわけではないです

> 問題2: 「ジャッカルの採用率が高いのを何とかしたい」

採用率が高いのはあまり問題ではないです。問題なのは役職のバランスを考えた時「ジャッカルとサイドキック」が前提になり始めてるのが問題です。(「Aという役職と「ジャッカルとサイドキック」を混ぜて使いたいけどBという部分でバランスが悪いんで、Aと「ジャッカルとサイドキック」を修正しよう！！」という流れが問題です)
採用率が高くなれば高くなるほどこの問題は顕著に出てき来ますが・・・

> ジャッカル/サイドキックをNerfする

バランスに問題があれば必要に応じてバランス調整は行います(多少対応コストが重くても対応はします)。現状バランスに問題があるとは思えないので調整は行いませんが
上げている例は私個人バランスに問題があると感じて無いけど要望がある感じです

> ジャッカルの勧誘周りを対応コストが低い形に修正する

どういう修正を行うかにもよりますとしか言えないです
能力を変えるというのであれば「ジャッカルとサイドキック」という役職とかけ離れているので、別役職対応になります

> 新たな勧誘能力を持ったニュートラル役職の作成

こちらもどんな勧誘能力を持つかにもよります

**Comment by melonjagaimo** [comment]
今しがた思いつきましたが、ジャッカル&サイドキックの逆で「親役職がキル能力を持たず、親によって生み出される子役職がキル能力を持つ」ニュートラルはどうですか？
子役職が親役職に昇格することはなく、親役職は死なないように動きながらキル能力持ちの子役職を作り、子役職は親役職と共に勝つべくキル能力を駆使するような

**Reply by yukieiji** [reply]
大本的には問題ないのですが、子役職を作った時の処理を考える必要がありますね
この役職の子役職作成処理は必ずこうであるって決めていれば問題ないと思います(サイドキック見たく中途半端に変化したり、変化しなかったり、作れないとかみたいなのがなければ問題はないと思いますよ)

**Reply by melonjagaimo** [reply]
親役職が非力な代わりに子役職作成は「必ず成功し、元々の役職を上書きする形で対象を子役職にする」と決めて考えてます

**Reply by yukieiji** [reply]
それなら問題ないと思います
バランス云々はチケットに上がってから考えますが

**Comment by yukieiji** [comment]
一瞬だけ頭の中に浮かんで必ず「ジャッカルとサイドキック」の変わりになる！！って思い文章にしたけど色々と問題が合ったため頓挫した役職案を置いておきす
問題点をうまい感じに解決できる方法を返信してくれると嬉しいです

- 名前未定
  - 勝利条件
    - 「工作」を100％にすること
  - 「工作員」
    - 能力で選定するが対象にも分からないし、役職を置き換えたり陣営を置き換えたりしない
    - 工作員が生存していてほかの生存者がいる場合、ゲームが終了しない（ここも推理要素ないなーって思う）
      - 例：クルー以外を全滅させたがゲームは終了しない(工作員を落とさないとゲームが終わらない)
  - 「工作」（問題点はここ、工作ゲージの上がり方がちょっと微妙）
    - キルをすることで工作ゲージが溜まっていく(これは問題なし)
    - 生きていると微増加
    - 工作ゲージは会議終了後に報告される(誤差付き)
    - 工作員の使用回数が0でこの役職が死亡した場合、会議終了後に正しい工作ゲージが表示され残りのゲージ分が工作員に割り当てられる(工作員が減れば減るほどゲージが大きく変動するためバレやすい)
    - 工作員の工作ゲージの進み
      - クルー：タスク
      - それ以外の陣営：それ以外なった瞬間 か タイマーで増加

**Reply by Nushigami** [reply]
自分が思いついた改善案として工作員の数を会議後に出すとかはどうでしょう?

**Reply by yukieiji** [reply]
工作員数は出してもいいとは思ってます

**Reply by melonjagaimo** [reply]
インポスター・ニュートラル陣営でもできて、推理の余地を残すやり方ですか。

1:工作員にしか見えない機械がマップの何処かに置かれていて、それに工作員が一定時間接近すると工作ゲージが増加する。

2:工作員が一定時間静止するとマップ上に機械を置いて工作ゲージ増加。ここで置かれた機械は全プレイヤーに見える。


というのはどうでしょうか

**Comment by SaidRed** [comment]
1に関して思い付いたことを

ロール優先度を全役職に設定して、優先度が高い役職はサイドキックのような役職変更を受け付けないようにするとかはどうでしょう。
例えばジャッカルとサイドキックは3でアサシンとマーリンは5と設定するとアサシンとマーリンはサイドキックにできない、のようなイメージです。

メリットは、今後追加の役職で変更されたら困る場合はデフォルトを高く設定するだけで良くなります。
それと、問題が分かった時に部屋主レベルでの対応もできるようになります。

**Reply by yukieiji** [reply]
今度はその優先度を毎役職ごとに考える必要が出てきます(またいま実装されてる50役職分考える + オプション値の実装とその翻訳)

**Comment by yukieiji** [comment]
クイーンがある程度機能していることを鑑みるに今後も増やす予定ではあるが優先度的には下がったと判断するためピンどめを外します
